### PR TITLE
feat(integrations): Create instrumented internal ApiClient and extend for Heroku

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
## Objective
We want to instrument the Heroku plugin. However this plugin makes requests to Sentry and we need to create a new subclass that instruments the current ApiClient with Datadog and Sentry Performance. Then we extend it with Heroku specific values.

### CODE REVIEW

File: src/sentry/shared_integrations/client.py [hunk 1]
```diff
@@ -310,3 +311,71 @@ def head_cached(self, path, *args, **kwargs):
...
+
+    @classproperty
+    def name_field(cls):
+        return u"%s_name" % cls.integration_type
+
+    @classproperty
+    def name(cls):
+        return getattr(cls, cls.name_field)
+
+    def track_response_data(self, code, span, error=None, resp=None):
```
Comment (scefali) [inline src/sentry/shared_integrations/client.py:335]: @NisanthanNanthakumar Can we make a mixin class for this and use it with `BaseApiClient`? No need to repeat code here.

File: src/sentry/shared_integrations/client.py [hunk 2]
```diff
@@ -171,6 +151,29 @@ def track_response_data(self, code, span, error=None, resp=None):
...
+
+    datadog_prefix = None
+
+    cache_time = 900
+
+    def __init__(self, verify_ssl=True, logging_context=None):
+        self.verify_ssl = verify_ssl
+        self.logging_context = logging_context
+
+    def get_cache_prefix(self):
```
Comment (scefali) [inline src/sentry/shared_integrations/client.py:174]: @NisanthanNanthakumar can we leave this in the mixin?

Comment (NisanthanNanthakumar) [reply src/sentry/shared_integrations/client.py:174]: @scefali we don't use that method in the `BaseInternalApiClient`. It's only used by the `BaseApiClient`. `get_cache_prefix`, `get_cached` and `head_cached` are only used by our first-party integrations. I don't think we should put those methods in the mixin bc they are not needed to for the `track_response_data` method.

File: src/sentry_plugins/heroku/plugin.py [hunk 3]
```diff
@@ -2,20 +2,29 @@
...
 from sentry.integrations import FeatureDescription, IntegrationFeatures
+from .client import HerokuApiClient
 
 logger = logging.getLogger("sentry.plugins.heroku")
 
 
 class HerokuReleaseHook(ReleaseHook):
+    def get_auth(self):
+        try:
+            auth = ApiKey(organization=self.project.organization, scope_list=["project:write"])
```
Comment (scefali) [inline src/sentry_plugins/heroku/plugin.py:20]: @NisanthanNanthakumar Nit: you can just return this line

**Review by scefali — APPROVED** [review]
Left one more comment but looks good

**Pull request merged**